### PR TITLE
[loadpath] Rework and simplify ML loadpath handling

### PR DESCRIPTION
**Pull request by ejgallego**
This PR refactors the handling of ML loadpaths to get it closer to
what (as of 2020) the standard OCaml toolchain (ocamlfind, dune) does.

This is motivated as I am leaning toward letting the standard OCaml
machinery handle OCaml includes; this has several benefits [for
example plugins become regular OCaml libs] It will also help in
improving dependency handling in plugin dynload.

The main change is that "recursive" ML loadpaths are no longer
supported, so Coq's `-I` option becomes closer to OCaml's semantics.

We still allow `-Q` to extend the OCaml path recursively, but this may
become deprecated in the future if we decide to install the ML parts
of plugins in the standard OCaml location.

Due to this `Loadpath` still hooks into `Mltop`, but other than that
`.v` location handling is actually very close to become fully
independent of Coq [thus it can be used in other tools such coqdep,
the build system, etc...]

In terms of vernaculars the changes are:

- The `Add Rec ML Path` command has been removed,
- The `Add Loadpath "foo".` has been removed. We now require that the
  form with the explicit prefix `Add Loadpath "foo" as Prefix.` is used.

We did modify `fake_ide` as not to add a directory with the empty
`Prefix`, which was not used. This exposed some bugs in the
implementation of the document model, which relied on having an
initial sentence; we have workarounded them just by adding a dummy one
in the two relevant cases.

**Kind:** infrastructure

- [x] Corresponding documentation was added / updated
- [x] Entry added in the changelog

### CODE REVIEW

**Review by herbelin — APPROVED** [review]
Trusting the motivation, though a bit worried about possible incompatibilities: are the removed commands used in the large?

**Comment by ejgallego** [comment]
> Trusting the motivation, though a bit worried about possible incompatibilities: are the removed commands used in the large?

Hard to know other than the CI, on the other hand their use case is pretty weak given our current infrastructure.

File: test-suite/ide/debug_ltac.fake [hunk 1]
```diff
@@ -1,2 +1,3 @@
+ADD { Definition foo := 0. }
```
Comment (SkySkimmer) [inline test-suite/ide/debug_ltac.fake]: Why this change?

File: stm/stm.ml [hunk 2]
```diff
@@ -2577,7 +2577,8 @@ type stm_init_options = {
 
   (* Initial load path in scope for the document. Usually extracted
      from -R options / _CoqProject *)
-  iload_path   : Loadpath.coq_path list;
+  ml_load_path   : CUnix.physical_path list;
+  vo_load_path   : Loadpath.vo_path list;
```
Comment (SkySkimmer) [inline stm/stm.ml]: Is the comment still up to date? I guess it should be changed to `Inital load paths` plural.

File: toplevel/coqargs.mli [hunk 3]
```diff
@@ -80,4 +80,4 @@ val parse_args : help:Usage.specific_usage -> init:t -> string list -> t * strin
 val error_wrong_arg : string -> unit
 
 val require_libs : t -> (string * string option * bool option) list
-val build_load_path : t -> Loadpath.coq_path list
+val build_load_path : t -> string list * Loadpath.vo_path list
```
Comment (SkySkimmer) [inline toplevel/coqargs.mli]:
```suggestion
val build_load_path : t -> CUnix.physical_path list * Loadpath.vo_path list
```
?

File: vernac/g_vernac.mlg [hunk 4]
```diff
@@ -1075,7 +1073,7 @@ GRAMMAR EXTEND Gram
     [ [ fl = LIST1 [ x = IDENT -> { x } ] -> { fl } ]]
   ;
   as_dirpath:
-    [ [ d = OPT [ "as"; d = dirpath -> { d } ] -> { d } ] ]
+    [ [ "as"; d = dirpath -> { d } ] ]
```
Comment (SkySkimmer) [inline vernac/g_vernac.mlg]: Might be nicer inlined.

File: vernac/vernacexpr.ml [hunk 5]
```diff
@@ -360,9 +359,9 @@ type nonrec vernac_expr =
   | VernacSolveExistential of int * constr_expr
 
   (* Auxiliary file and library management *)
-  | VernacAddLoadPath of rec_flag * string * DirPath.t option
+  | VernacAddLoadPath of bool * string * DirPath.t
```
Comment (SkySkimmer) [inline vernac/vernacexpr.ml]: Maybe worth turning into an inline record?

**Review by SkySkimmer — APPROVED** [review]

File: doc/changelog/11-infrastructure-and-dependencies/11618-loadpath+split_ml_handling.rst [hunk 6]
```diff
@@ -0,0 +1,7 @@
+- **Changed:**
+  Recursive OCaml loadpaths are not supported anymore; the command
+  ``Add Rec ML Path`` has been removed. We have also dropped support
+  for the non-qualified version of the ``Add Loadpath dir`` command,
```
Comment (Zimmi48) [inline doc/changelog/11-infrastructure-and-dependencies/11618-loadpath+split_ml_handling.rst]:
```suggestion
  for the non-qualified version of the :cmd:`Add Loadpath` command,
```

File: doc/sphinx/proof-engine/vernacular-commands.rst [hunk 7]
```diff
@@ -757,11 +757,6 @@ the toplevel, and using them in source files is discouraged.
    :n:`-R @string @dirpath`. It adds the physical directory string and all its
    subdirectories to the current |Coq| loadpath.
 
-   .. cmdv:: Add Rec LoadPath @string
```
Comment (Zimmi48) [inline doc/sphinx/proof-engine/vernacular-commands.rst]: You have removed this variant but not the `Add LoadPath @string` variant. Is this a mistake?

File: test-suite/ide/debug_ltac.fake (hunk 1 above)
Comment (ejgallego) [reply test-suite/ide/debug_ltac.fake]: As written in the commit message:

> We did modify `fake_ide` as not to add a directory with the empty
`Prefix`, which was not used. This exposed some bugs in the
implementation of the document model, which relied on having an
initial sentence; we have workarounded them just by adding a dummy one
in the two relevant cases.

A better solution would be welcome but it is not trivial.

File: stm/stm.ml (hunk 2 above)
Comment (ejgallego) [reply stm/stm.ml]: I've updated the comments and ported them to ocamldoc.

File: toplevel/coqargs.mli (hunk 3 above)
Comment (ejgallego) [reply toplevel/coqargs.mli]: Done, thanks. [By the way I tried at some point to add a `Path.t` opaque type to Coq, it was not so easy as I expected]

File: vernac/vernacexpr.ml (hunk 5 above)
Comment (ejgallego) [reply vernac/vernacexpr.ml]: Done thanks.

File: doc/changelog/11-infrastructure-and-dependencies/11618-loadpath+split_ml_handling.rst (hunk 6 above)
Comment (ejgallego) [reply doc/changelog/11-infrastructure-and-dependencies/11618-loadpath+split_ml_handling.rst]: thanks!

File: vernac/g_vernac.mlg (hunk 4 above)
Comment (ejgallego) [reply vernac/g_vernac.mlg]: Done thanks.

File: doc/sphinx/proof-engine/vernacular-commands.rst (hunk 7 above)
Comment (ejgallego) [reply doc/sphinx/proof-engine/vernacular-commands.rst]: Ups, I was under the impression I had effectively taken care of both of them.

**Comment by ejgallego** [comment]
This is ready, assignee would be great @coq/build-maintainers is the right team for this IMO.

**Comment by Zimmi48** [comment]
@SkySkimmer You've reviewed the code. Do you want to assign?

**Comment by Zimmi48** [comment]
I can also self-assign myself, but then I will have do a more thorough review myself which won't be for another couple of days.

File: doc/changelog/11-infrastructure-and-dependencies/11618-loadpath+split_ml_handling.rst [hunk 8]
```diff
@@ -0,0 +1,9 @@
+- **Changed:**
```
Comment (Zimmi48) [inline doc/changelog/11-infrastructure-and-dependencies/11618-loadpath+split_ml_handling.rst]: It seems to me that this should be "Removed" since all the changelog entry is about removal.

File: test-suite/ide/debug_ltac.fake (hunk 1 above)
Comment (Zimmi48) [reply test-suite/ide/debug_ltac.fake]: It seems like `.fake` files support comments (with `#`). Would you mind adding one explaining this so that it is clear why it is like that in the future?

File: test-suite/ide/undo002.fake [hunk 9]
```diff
@@ -3,6 +3,7 @@
 #
 # Simple backtrack by 2 before two global definitions
 #
+ADD initial { Definition moo := 3. }
```
Comment (Zimmi48) [inline test-suite/ide/undo002.fake]: Same thing here.

**Review by Zimmi48 — COMMENTED** [review]
I'll assign this PR because it has been waiting for way too long (sorry @ejgallego). But @gares, would you mind having a quick look at the changes to the STM / fakeide? Thanks!

**Comment by gares** [comment]
I think this is ok. The fakeide thing is not really a bug, it is just that before this patch all documents had an implicit initial sentence setting the loadpath. Now they don't have it, and some tests had to just back to that initial sentence. So Emilio added a dummy sentence. I think I'd prefer that initial sentence to be a no-op one, like `Check Prop`, rather than defining `foo`, that in once case cannot be called foo because there is already a foo... Of course adding a comment in the .fake files would be nice.

File: test-suite/ide/debug_ltac.fake (hunk 1 above)
Comment (ejgallego) [reply test-suite/ide/debug_ltac.fake]: I've used a `Comments` command.

File: doc/changelog/11-infrastructure-and-dependencies/11618-loadpath+split_ml_handling.rst (hunk 8 above)
Comment (ejgallego) [reply doc/changelog/11-infrastructure-and-dependencies/11618-loadpath+split_ml_handling.rst]: I used changed as we changed the semantics of includes, I dunno, I see this more as a change on how loadpaths are handled [same now than in OCaml] vs a real removal, maybe we could add both?

**Comment by ejgallego** [comment]
Thanks folks, I've used the `Comments` venacular to both provide and document the initial sentences in `.fake` tests.

File: doc/changelog/11-infrastructure-and-dependencies/11618-loadpath+split_ml_handling.rst (hunk 8 above)
Comment (Zimmi48) [reply doc/changelog/11-infrastructure-and-dependencies/11618-loadpath+split_ml_handling.rst]: It's fine to keep Changed if you think this is more appropriate.

File: vernac/vernacexpr.ml [hunk 10]
```diff
@@ -363,9 +362,13 @@ type nonrec vernac_expr =
   | VernacSolveExistential of int * constr_expr
 
   (* Auxiliary file and library management *)
-  | VernacAddLoadPath of rec_flag * string * DirPath.t option
+  | VernacAddLoadPath of { implicit : bool
```
Comment (maximedenes) [inline vernac/vernacexpr.ml]: Is `implicit` the right name here? I seem to recall it was used somewhere else in the codebase but was a bit misleading. What does it mean here?

Comment (maximedenes) [reply vernac/vernacexpr.ml]: Sorry, I missed it above. So, shouldn't it be `recursive` instead?

Comment (maximedenes) [reply vernac/vernacexpr.ml]: Ah no, I see what the patch is doing: previously `isrec` was passed to `implicit`. This patch distinguishes the two terminologies more clearly.

**Review by maximedenes — APPROVED** [review]
LGTM
